cpp/throwaways/check_archive_org_isbn_presence.cc

```diff
@@ -0,0 +1,133 @@
+#include <mutex>
+#include <thread>
+#include <cstdio>
+#include <cstdlib>
+#include "Downloader.h"
+#include "FileUtil.h"
+#include "StringUtil.h"
+#include "MARC.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("[--summarize-tags] [--verbose] worker_thread_count marc_data isbn_list_output");
+}
+
+
+std::atomic<bool> work_available(true);
+
+
+void WorkerThread(Downloader * const downloader, std::deque<std::set<std::string>> * const task_queue,
+                  std::mutex * const task_queue_mutex, unsigned * const isbn_found_count, File * const isbn_list_output,
+                  std::mutex * const output_mutex)
+{
+    std::set<std::string> isbns;
+    while (work_available) {
+        {
+            std::lock_guard<std::mutex> task_queue_mutex_locker(*task_queue_mutex);
+            if (not task_queue->empty()) {
+                isbns = task_queue->front();
+                task_queue->pop_front();
+            }
+        }
+
+        if (isbns.empty()) {
+            std::this_thread::sleep_for(std::chrono::seconds(20));
+            continue;
+        }
+
+        for (const auto &isbn : isbns) {
+            const std::string url("https://archive.org/metadata/isbn_" + isbn + "/created");
+            if (not downloader->newUrl(url)) {
+                LOG_WARNING("URL \"" + url + " failed to download!");
+                continue;
+            }
+
+            if (downloader->getMessageBody().find("result") != std::string::npos) {
+                std::lock_guard<std::mutex> output_mutex_locker(*output_mutex);
+                isbn_list_output->writeln(isbn);
+                ++*isbn_found_count;
+                std::cout << *isbn_found_count << '\n';
+                break;
+            }
+        }
+
+        isbns.clear();
+    }
+}
+
+
+void ProcessRecords(MARC::Reader * const marc_reader, std::deque<std::set<std::string>> * const task_queue,
+                    std::mutex * const task_queue_mutex, unsigned * const isbn_found_count)
+{
+    Downloader downloader;
+    unsigned record_count(0), monographs_with_isbns_found_count(0);
+    while (const MARC::Record record = marc_reader->read()) {
+        ++record_count;
+        const auto isbns(record.getISBNs());
+        if (isbns.empty())
+            continue;
+
+        ++monographs_with_isbns_found_count;
+
+        std::lock_guard<std::mutex> task_queue_mutex_locker(*task_queue_mutex);
+        task_queue->push_back(isbns);
+    }
+    work_available = false; // Let our worker threads return.
+
+    LOG_INFO("Processed " + std::to_string(record_count) + " record(s) and found "
+             + std::to_string(*isbn_found_count) + " ISBN('s) that are/is present on archive.org.");
+}
+
+
+} // unnamed namespace
+
+
+int Main(int argc, char *argv[]) {
+    if (argc != 4)
+        Usage();
+
+    const unsigned WORKER_THREAD_COUNT(StringUtil::ToUnsigned(argv[1]));
+    std::vector<std::thread> thread_pool((WORKER_THREAD_COUNT));
```

Comment:
double (()) truly needed here?

Reply:
Yes, this is because of the "most vexing parse": https://en.wikipedia.org/wiki/Most_vexing_parse